charmcraft/config.py

```diff
@@ -18,167 +18,195 @@
-    storage_url = attr.ib(default="https://storage.snapcraftcontent.com")
+    loc = ".".join(loc_parts)
+
+    # Filter out internal __root__ detail.
+    loc = loc.replace(".__root__", "")
+    return loc
+
+
+def format_pydantic_error_msg(msg):
+    """Format pydantic's error message field."""
+    # Replace shorthand "str" with "string".
+    msg = msg.replace("str type expected", "string type expected")
+    return msg
 
-    @classmethod
-    def from_dict(cls, source):
-        """Build from a raw dict."""
-        return cls(**source)
 
+def format_pydantic_errors(errors):
+    """Format errors.
 
-class BasicPrime(tuple):
-    """Hold the list of files to include, specified under parts/bundle/prime configs.
+    Example 1: Single error.
 
-    This is a intermediate structure until we have the full Lifecycle in place.
+    Bad charmcraft.yaml content:
+    - field: <some field>
+      reason: <some reason>
+
+    Example 2: Multiple errors.
+
+    Bad charmcraft.yaml content:
+    - field: <some field>
+      reason: <some reason>
+    - field: <some field 2>
+      reason: <some reason 2>
     """
+    combined = ["Bad charmcraft.yaml content:"]
+    for error in errors:
+        loc = "- field: " + format_pydantic_error_location(error["loc"])
+        combined.append(loc)
 
-    @classmethod
-    def from_dict(cls, parts):
-        """Build from a dicts sequence."""
-        prime = parts.get("bundle", {}).get("prime", [])
-        return cls(prime)
+        msg = "  reason: " + format_pydantic_error_msg(error["msg"])
+        combined.append(msg)
+
+    return "\n".join(combined)
+
+
+class Part(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Placeholder representation for part.  To be replaced by craft-parts."""
+
+    prime: List[pydantic.StrictStr] = []
+
+    @pydantic.validator("prime", each_item=True)
+    def validate_relative_paths(cls, prime):
+        """Verify relative paths are used in prime."""
+        return check_relative_paths(prime)
+
+
+class Parts(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Represents parts: top-level key.  To be replaced by craft-parts."""
+
+    __root__: Dict[pydantic.StrictStr, Part] = {}
 
+    def get(self, part_name) -> Part:
+        """Get part by name.
 
-@attr.s(kw_only=True, frozen=True)
-class Project:
+        :returns: Part if exists, None if not.
+        """
+        try:
+            return self.__root__[part_name]
+        except KeyError:
+            return None
+
+
+class CharmhubConfig(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Represents top-level charmhub object."""
+
+    api_url: pydantic.StrictStr = "https://api.charmhub.io"
+    storage_url: pydantic.StrictStr = "https://storage.snapcraftcontent.com"
+
+    @pydantic.validator("api_url", "storage_url")
+    def validate_urls(cls, url):
```

Comment:
Can be removed if validating using `HttpUrl`.

Reply:
Nice, didn't notice that one.

Reply:
Done!